packages/pwa-kit-react-sdk/src/ssr/universal/hooks/use-server-effect.tsx

```diff
@@ -0,0 +1,215 @@
+    */
+    ServerEffectProvider: Provider; // Can we get more specifc here? provider type?
+    /**
+    * useServerEffect: A useEffect style hook that can be used client and server side to run async
+    * actions.
+    */
+    useServerEffect: ServerEffect;
+    /**
+    * resolveEffects: For use on the server side only, this async function will can be called to invoke
+    * all the recorded async "didUpdate" functions.
+    */
+    resolveEffects: EffectsResolver;
+}
+type ServerEffectContextValue = {
+    name: string;
+    data: object;
+    requests: DidUpdateFn[];
+    resolved: boolean;
+}
+
+/**
+ * This function is creating `useServerEffect`, it's only going to be used
+ * but the SDK, this isn't for external libs.
+ * 
+ * @param {*} context 
+ * @returns 
+ */
+const createServerEffect = (context) => {
+    const useServerEffect = (didUpdate: DidUpdateFn, source: any[], initial?: any) : ServerEffectState => {  
+        debugger  
+        const key = useUID()
+        const ctx: ServerEffectContextValue = useContext(context)
+
+        const getPropsArgs = useGetPropsArgs()
+        const [data, setData] = useState(ctx.data[key] || initial)
+        const [loading, setLoading] = useState(false)
+        const [error, setError] = useState(undefined)
+        const boundDidUpdate = didUpdate.bind(this, getPropsArgs)
+
+        const wrappedDidUpdate = isServer ? 
+            () => {} : 
+            async () => {
+                // Set loading state to "true"
+                setLoading(true)
+        
+                try {
+                    const data = await boundDidUpdate()
+                    setData(data)
+                } catch(e) {
+                    setLoading(false)
+                    setError(e)
+                }
+        
+                // Set loading state to false
+                setLoading(false)
+            }
+    
+        console.log('PUSHING REQUEST: ', !ctx.resolved, isServer)
+        if (!ctx.resolved && isServer) {
+            
+            ctx.requests.push(async () => {
+                const data = await boundDidUpdate()
+    
+                return {
+                    [key]: data
+                }
+            })
+        }
+    
+        // To avoid messing with React's hooks order lets make this a
+        // noop for the server.
+        useEffect(wrappedDidUpdate, source)
+    
+        return {
+            loading,
+            error,
+            data
+        }
+    }
+
+    return (...args) => (
+        useServerEffect.apply(this, args)
+    )
+}
+
+/**
+ * Create a provider for use with hooks created with the paired createSeverEffect function.
+ *
+ * @param {string} name - for namspacing and data accessing purposes.
+ * 
+ * @function
+ * 
+ * @returns {ServerEffectProvider}
+ */
+export const createServerEffectContext = (name: string): ServerEffectContext => {
+    console.log('--=-=-=---=-=-=-=---==--==-> createServerEffectContext')
+    const ctxValue = {
```

Comment:
```suggestion
    const ctxValue: WhateverContextValueType = {
```
Without specifying a type, `requests` defaults to `never[]`, which is not what we want.